src/core/lib/security/transport/auth_filters.h

```diff
@@ -51,6 +52,93 @@ class ClientAuthFilter final : public ChannelFilter {
   static absl::StatusOr<std::unique_ptr<ClientAuthFilter>> Create(
       const ChannelArgs& args, ChannelFilter::Args);
 
+ private:
+  auto GetMetadataFromCreds(RefCountedPtr<grpc_call_credentials> creds,
+                            ClientMetadataHandle md) {
+    return Map(creds->GetRequestMetadata(std::move(md), &args_),
+               [](absl::StatusOr<ClientMetadataHandle> new_metadata) mutable {
+                 if (!new_metadata.ok()) {
+                   return absl::StatusOr<ClientMetadataHandle>(
+                       MaybeRewriteIllegalStatusCode(new_metadata.status(),
+                                                     "call credentials"));
+                 }
+                 return new_metadata;
+               });
+  }
+
+  // Early in the declaration order as auto return type is needed by Call.
+  auto GetCallCredsMetadata(ClientMetadataHandle md) {
+    auto creds = GetCallCreds(*md);
+    return If(
+        creds.ok(),
+        [this, &creds, md = std::move(md)]() mutable {
+          return If(
+              *creds != nullptr,
+              [this, &creds, &md]() {
+                return GetMetadataFromCreds(std::move(*creds), std::move(md));
+              },
+              [&md]() {
+                return Immediate(
+                    absl::StatusOr<ClientMetadataHandle>(std::move(md)));
+              });
+        },
+        [&creds]() {
+          return Immediate(
+              absl::StatusOr<ClientMetadataHandle>(creds.status()));
+        });
+  }
+
+ public:
+  class Call {
+   public:
+    auto OnClientInitialMetadata(ClientMetadataHandle md,
+                                 ClientAuthFilter* filter) {
+      filter->InstallContext();
+      auto* host = md->get_pointer(HttpAuthorityMetadata());
+      return WithResult<absl::StatusOr<ClientMetadataHandle>>(If(
```

Comment:
Why do we need `WithResult<>` here?

Reply:
it's a zero-cost debug helper, but woah it helped on this one

this is basically "assert that the resolved value of this promise is as expected" and I had a bug in the if that was throwing it off -- given i spent an hour on the fix before adding this, I left it in place